internal/extproc/dynlb/dynlb.go

```diff
@@ -0,0 +1,171 @@
+		logger: logger,
+		models: make(map[string]filterapi.DynamicLoadBalancingModel, len(dyn.Models)),
+	}
+
+	// TODO: maybe reuse the client for multiple queries.
+	client := dns.Client{}
+	conn, err := client.Dial(dnsServerAddr)
+	if err != nil {
+		return nil, fmt.Errorf("failed to dial DNS server: %w", err)
+	}
+	defer conn.Close()
+	for _, b := range dyn.Backends {
+		for _, ip := range b.IPs {
+			ret.endpoints = append(ret.endpoints, endpoint{
+				ipPort:  []byte(fmt.Sprintf("%s:%d", ip, b.Port)),
+				backend: &b.Backend,
+			})
+		}
+		logger.Info("resolving hostnames to IP addresses", slog.String("hostnames", strings.Join(b.Hostnames, ",")))
+		// Resolves all hostnames to IP addresses.
+		for _, hostname := range b.Hostnames {
+			// Append a dot if the hostname is not fully qualified.
+			fqdn := hostname
+			if !strings.HasSuffix(fqdn, ".") {
+				fqdn += "."
+			}
+			msg := new(dns.Msg)
+			// TODO: add support for TypeAAAA for IPv6.
+			msg.SetQuestion(fqdn, dns.TypeA)
+			response, _, err := client.ExchangeWithConnContext(ctx, msg, conn)
+			if err != nil {
+				return nil, fmt.Errorf("failed to query DNS server: %w", err)
+			}
+			if response.Rcode != dns.RcodeSuccess {
+				return nil, fmt.Errorf("DNS query failed: %s", dns.RcodeToString[response.Rcode])
+			}
+
+			for _, answer := range response.Answer {
+				if aRecord, ok := answer.(*dns.A); ok {
+					logger.Info("resolved IP address", slog.String("hostname", hostname), slog.String("ip", aRecord.A.String()))
+					ret.endpoints = append(ret.endpoints, endpoint{
+						ipPort:   []byte(fmt.Sprintf("%s:%d", aRecord.A.String(), b.Port)),
+						backend:  &b.Backend,
+						hostname: hostname,
+					})
+				}
+			}
+		}
+	}
+	for _, m := range dyn.Models {
+		ret.models[m.Name] = m
+	}
+	return ret, nil
+}
+
+// dynamicLoadBalancer implements DynamicLoadBalancer.
+type dynamicLoadBalancer struct {
+	logger    *slog.Logger
+	models    map[string]filterapi.DynamicLoadBalancingModel
+	endpoints []endpoint
+}
+
+// endpoint represents an endpoint, a pair of IP and port, which belongs to a backend.
+type endpoint struct {
+	ipPort []byte
+	// hostname is the hostname used to resolve the IP address. Can be empty if the IP is not resolved from a hostname.
+	hostname string
+	// backend is the backend that this ip:port pair belongs to.
+	backend *filterapi.Backend
+}
+
+// SelectChatCompletionsEndpoint implements [DynamicLoadBalancer.SelectChatCompletionsEndpoint].
+//
+// TODO: expand x.ChatCompletionMetrics to add getter methods to be able to make a decision based on the metrics.
+// TODO: this might need to return dynamic metadata instead of headers.
+func (dlb *dynamicLoadBalancer) SelectChatCompletionsEndpoint(model string, _ x.ChatCompletionMetrics) (
+	selected *filterapi.Backend, headers []*corev3.HeaderValueOption, err error,
+) {
+	m, ok := dlb.models[model]
+	if !ok {
+		err = fmt.Errorf("model %s is not found in the dynamic load balancer", model)
```

Comment:
This may not be necessarily an error as `models` is optional in dlb, it could a list of backend hosts.